### PR TITLE
Substrate front-end tutorial

### PR DESCRIPTION
**Pull request by Tbaut**
This tutorial is meant for UI developers and should not require anything else than basic React JS knowledge. Someone not familiar with React hooks should be able to understand how they work without having to read anything else.

not to do (for other tutorials)
- teach about the whole srml -> metadata -> Polkadot-js api workflow
- teach how to inject custom types
- teach how to access custom storage/extrinsics
- teach how to do a runtime upgrade

### CODE REVIEW

**Comment by jeluard** [comment]
How can I browse the end result? Should it be included in the tutorials section?

**Comment by Tbaut** [comment]
Each section has something like

>You can get the working version of this code by visiting `part-1-1` directory and launch it with:
```bash
cd part-1-1;
yarn;
yarn start;
```

And the [basic-dapp repo](https://github.com/substrate-developer-hub/basic-dapp) has a succint readme (also for each part).

File: docs/tutorials/basic-dapp/part-1-connect-to-a-node.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+      setNodeInfo ({
+        chain,
+        nodeName,
+        nodeVersion
+      })
+    })
+    .catch((e) => console.error(e));
+  },[api.rpc.system]);
+```
+Remember what we said in the previous part? The second argument of `useEffect` is an array that presents the variables to watch. In this case, `useEffect` will be triggered when the component loads and whenever `api.rpc.system` changes. It should not but we never know (and if you don't specify this argument the component will work surely as well but you'll see a warning in the console).
```
Comment (Stefie) [inline docs/tutorials/basic-dapp/part-1-connect-to-a-node.md]:
```suggestion
Remember what we said in the previous part? The second argument of `useEffect` is an array that presents the variables to watch. In this case, `useEffect` will be triggered when the component loads and whenever `api.rpc.system` changes.
The API endpoints do not change unless there's a runtime upgrade of the node. Even though this will not happen in our development environment and the component will work without specifying this argument, we are still passing the `api.rpc.system` to `useEffect` to make our examples as realistic as possible.
```

File: docs/tutorials/basic-dapp/part-1-connect-to-a-node.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+Now that we are successfully connected to our node let's present some information about it.
+We will create a new component called `NodeInfo` responsible for displaying this information, and we will pass the `api` in the props.
+All we need to import are the usual `React` component and the hook functions.
+
+```js
+import React, {useEffect, useState} from 'react';
+```
+
+The [Polkadot-js api](https://polkadot.js.org/api/api/) allows you to send RPC queries to the node.
```
Comment (Stefie) [inline docs/tutorials/basic-dapp/part-1-connect-to-a-node.md]:
```suggestion
The [Polkadot-js api](https://polkadot.js.org/api/api/) allows you to send RPC calls to the node.
```

File: docs/tutorials/basic-dapp/part-1-connect-to-a-node.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+To get the node's information we want, we will use the `rpc.system` method and section given by the `api` promise.
```
Comment (Stefie) [inline docs/tutorials/basic-dapp/part-1-connect-to-a-node.md]:
```suggestion
To get the node's information we want, we will use the `rpc.system` methods provided by the `api`.
```

File: docs/tutorials/basic-dapp/part-1-connect-to-a-node.md [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+We will persist in the state the information returned as an object, to display it later on.
```
Comment (Stefie) [inline docs/tutorials/basic-dapp/part-1-connect-to-a-node.md]:
```suggestion
We will add the information returned by the API to the components state to be able to access and display it later on.
```

File: docs/tutorials/basic-dapp/part-1-connect-to-a-node.md [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        nodeName,
+        nodeVersion
+      })
+    })
+    .catch((e) => console.error(e));
+  },[api.rpc.system]);
+```
+Remember what we said in the previous part? The second argument of `useEffect` is an array that presents the variables to watch. In this case, `useEffect` will be triggered when the component loads and whenever `api.rpc.system` changes. It should not but we never know (and if you don't specify this argument the component will work surely as well but you'll see a warning in the console).
+
+Finally, we'll display these info at the top of the page:
```
Comment (Stefie) [inline docs/tutorials/basic-dapp/part-1-connect-to-a-node.md]:
```suggestion
Finally, we'll display our node information at the top of the page:
```

File: docs/tutorials/basic-dapp/index.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
+---
+title: "Introduction"
+---
+
+This is a beginner level guide to help you get started with building a basic decentralized application (later called DApp) on top of the Substrate based blockchain, using [polkadot{.js} api (aka polkadot-js api)](https://github.com/polkadot-js). This tutorial does not require any previous knowledge about the Substrate framework or rust language, it does, however, require knowledge of Javascript as well a basic knowledge of the React framework.
```
[the remark this answers is not in the document]
Comment (Tbaut) [reply docs/tutorials/basic-dapp/index.md]: Right, the {.js} comes from the repo, https://github.com/polkadot-js/ also we use it in the extension's name. It look good but that's not very search engine friendly.

[the remark this answers is not in the document]
Comment (Tbaut) [reply docs/tutorials/basic-dapp/index.md]: I went for "Polkadot-js api" because we usually have anything after Polkadot-js in lower case such as `apps`, `extension`, `crypto`...

File: docs/tutorials/basic-dapp/part-3-transfer-funds.md [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+```
+
+You can get the working version of this code by visiting `part-3-1` directory:
+
+```bash
+cd part-3-1;
+yarn;
+yarn start;
+```
+If you run this example, you will get find our `Transfert` component below the Balance table.
```
Comment (Stefie) [inline docs/tutorials/basic-dapp/part-3-transfer-funds.md]:
```suggestion
If you run this example, you will get find our `Transfer` component below the Balance table.
```

File: docs/tutorials/basic-dapp/part-3-transfer-funds.md [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    setFormState(formState => {
+      return {
+        ...formState,
+        [data.state]: data.value
+      };
+    });
+  };
+```
+
+We will then create our function to actually submit the form. To do so we need to create the transaction, then sign it with our private key and finally send it. The `api` provides all we need here again, with `api.tx.balances` and the `transfer` and `signAndSend` methods. The latter takes the Keypair as the first parameter and a callback as the second parameter, passing to the callback the `status` of the transaction that we will use to see if our transaction is executed successfully thanks to the `isFinalized` flag. In the example below, we will show our users if the transaction is successful or not and in case it is successful at what block it was executed.
```
Comment (Stefie) [inline docs/tutorials/basic-dapp/part-3-transfer-funds.md]:
```suggestion
We will then create our function to actually submit the form. To do so we need to create the transaction, then sign it with our private key and finally send it. The `api` provides all we need here again, with `api.tx.balances` and the `transfer` and `signAndSend` methods. The latter takes the Keypair as the first parameter and a callback as the second parameter.
The the `status` of the transaction is passed to the callback function. We will use it to see if our transaction is executed successfully. In the example below, we will show our users if the transaction was successfully finalized using the `isFinalized` status value. If it was successful executed, we're showing at which block it was finalized.
```

**Comment by Stefie** [comment]
Second batch of comments. It might actually be nice to have a native speaker looking over it! I'm not 100% sure some of the grammar stuff

File: docs/tutorials/basic-dapp/part-2-query-balances.md [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+  }, [accounts, api.query.balances, api.query.balances.freeBalance]);
+```
+
+We start by creating a `balance` state, that we will initialize at `0`. Then, as we did in the previous part, we will make the query in a React hook, with `useEffect`. However, we will want to display if the balance changes, this `api.query.balances.freeBalance` takes as first argument the account's address you want to request, and you can pass a callback as the second argument. This callback receives the current balance for the given account. This is where we will persist it in the state.
+
+This `query` is a subscription to the storage, this is what will allow us to see the balance change without reloading the page! Because it's a subscription, we must unsubscribe whenever our `Balances` React component unmounts. The `unsubscribe` function is what is actually returned by the `query`. This one must be called in the `return` statement of the `useEffect` that will be called whenever the component unmounts, in our case, it's when we close the tab or refresh out app's page manually.
+
+Let's strip down the table we return to only show Alice and the account's associated balance.
```
[the remark this answers is not in the document]
Comment (Tbaut) [reply docs/tutorials/basic-dapp/part-2-query-balances.md]: I think this one is right, although not nice. I'll change the formulation.

File: docs/tutorials/basic-dapp/part-2-query-balances.md [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+```
+
+If you run this example, you will get a table with "Alice" account and balance.
+![Alice balance](/docs/tutorials/basic-dapp/assets/part-2-2.jpg)
+
+## 2.3 Query and display all balances
+
+Now that we know how querying the state of a module works, we can map through all of our accounts and query the state for each one. The `api` has a handy function to batch multiple queries at the same time, it's called `multi`. You can pass an array of arguments and it will return an array of the result.
+
+Instead of having just the balance of Alice in the state, we will change it an object mapping `address` with `freeBalance` so that we can use it to display each account's balance. Because the `multi` remains one subscription, the logic for unsubscribing remains the same.
```
[the remark this answers is not in the document]
Comment (Tbaut) [reply docs/tutorials/basic-dapp/part-2-query-balances.md]: I think that one is right (the "acocunt's") too.

File: docs/tutorials/basic-dapp/part-2-query-balances.md [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+
+    return () => unsubscribe && unsubscribe();
+  }, [accounts, api.query.balances, api.query.balances.freeBalance]);
+```
+
+We start by creating a state variable called `balance`, that we will initialize at `0`. Then, as we did in the previous part, we will make the query in a React hook with `useEffect`. 
+We want to display the balance and subscribe to all changes populated by the Substrate node.
+
+**Subscribe**
+The API method `api.query.balances.freeBalance()` takes  an accounts address as its first argument and you can pass an optional callback as the second argument to subscribe to any updates. The callback function is then listening to changes of the `freeBalance` value  of the given account.
```
Comment (Tbaut) [inline docs/tutorials/basic-dapp/part-2-query-balances.md]:
```suggestion
The API method `api.query.balances.freeBalance()` takes  an account address as its first argument and you can pass an optional callback as the second argument to subscribe to any update. The callback function is then listening to changes of the `freeBalance` value  of the given account.
```

File: docs/tutorials/basic-dapp/part-2-query-balances.md [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+
+**Subscribe**
+The API method `api.query.balances.freeBalance()` takes  an account address as its first argument and you can pass an optional callback as the second argument to subscribe to any update. The callback function is then listening to changes of the `freeBalance` value  of the given account.
+
+We're updating the components state with the values we're getting from the API and that's what allows us to see the balance change without reloading the page! 
+
+**Unsubscribe**
+But because it's a subscription, we must unsubscribe whenever our `Balances` React component unmounts. The `unsubscribe` function is what is actually returned by the `query`. This one must be called in the `return` statement of the `useEffect` that will be called whenever the component unmounts. In our case, it's when we close the tab or refresh our applications manually.
+
+Let's strip down the table we return to only show Alice and the account's associated balance.
```
Comment (Tbaut) [inline docs/tutorials/basic-dapp/part-2-query-balances.md]:
```suggestion
Let's strip down the table we return to only show the balance for Alice.
```

File: docs/tutorials/basic-dapp/part-4-use-external-accounts.md [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+Create an account using the extension and give it a name.
+
+We will now adapt our application to inject any account created externally in our current interface.
+
+This is how the extension looks like with 2 accounts creatively named Bob and Alice:
+![Pokadot-js extension](/docs/tutorials/basic-dapp/assets/part-3-1.jpg)
+
+## 4.2 Display external account's balances
+
+Externally generated accounts loaded thank to [`@polkadot/extension-dapp`](https://github.com/polkadot-js/extension/tree/master/packages/extension-dapp).
```
Comment (Stefie) [inline docs/tutorials/basic-dapp/part-4-use-external-accounts.md]: I don't really understand this sentence. There seems to be something missing...

File: docs/tutorials/basic-dapp/part-4-use-external-accounts.md [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+```bash
+yarn add @polkadot/extension-dapp@beta
+```
+
+This package gives us access to `web3Enable` to declare our DApp and request the right to read the accounts from our users. `isWeb3Injected` lets us know if our users have an extension supporting the `@polkadot/extension-dapp` protocol, and finally the `web3Accounts` function that should be called if the user has a compatible extension, and has granted us access to their accounts.
+
+To keep things simple, we will not display anything else than a loader to our users while the authorization is requested. This is a blocking experience that should not be reproduced in a real DApp. Rather show something to your users, and patiently wait for them to accept the extension request.
+
+We will introduce a new state variable `accountLoaded` to show this loader in case our user have an extension but haven't granted us access to their accounts.
```
Comment (Stefie) [inline docs/tutorials/basic-dapp/part-4-use-external-accounts.md]:
```suggestion
We will introduce a new state variable `accountLoaded` to show this loader in case our user has an extension but hasn't granted us access to their accounts.
```

File: docs/tutorials/basic-dapp/part-4-use-external-accounts.md [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+
+```js
+// Toward the top of Apps.js
```
Comment (Stefie) [inline docs/tutorials/basic-dapp/part-4-use-external-accounts.md]: I'd rephrase that somehow.

File: docs/tutorials/basic-dapp/part-4-use-external-accounts.md [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+const [accountLoaded, setaccountLoaded] = useState(false);
+
+// in our render function, show a loader to users with a compatible extension while 
+// waiting for their authorization
+if (!accountLoaded) {
+return loader('Loading accounts (please review any extension\'s authorization)');
+}
+```
+
+We will declare our DApp and request access to the external accounts in a `useEffect` React hook that we are now familiar with. We used to call `loadAccounts()` to initialize the keyring. This function also accepts injected accounts as arguments, so all we'll have to do is `loadAccounts(injectedAccounts)`. Quite handy isn't it?
```
Comment (Stefie) [inline docs/tutorials/basic-dapp/part-4-use-external-accounts.md]: Starting a war on commas here. I'd remove all of the ones before `"and"` and add one here for example `"Quite handy, isn't it"`, but I'm not really familiar with the english comma rules. That would just be me adapting the german rules to english

File: docs/tutorials/basic-dapp/index.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (Stefie) [reply docs/tutorials/basic-dapp/index.md]: The thing is, that both  'JS' and 'API' are fixed terms and both are correctly written in uppercase. Opposed to that, apps, extension etc. are correctly written in lowercase (unless used as a name). 
That's why  "Polkadot-js api" just looks really wrong to me. 

For JS see:
https://developer.mozilla.org/en-US/docs/Web/JavaScript
https://js.foundation/

And for API, just google 'api' orhttps://developer.mozilla.org/en-US/docs/Web/API

**Comment by jeluard** [comment]
Would it make sense to agree on some structure for tutorials split in steps? e.g. collectables one has steps too, but they have a fairly differently layout (e.g. `5/assets/5.1-finished-code.rs`)

cc @shawntabrizi

**Comment by shawntabrizi** [comment]
> Would it make sense to agree on some structure for tutorials split in steps? e.g. collectables one has steps too, but they have a fairly differently layout (e.g. `5/assets/5.1-finished-code.rs`)

Hmm, dunno if we need so much structure at this point?

I would like an end goal of anyone in the community writing and sharing tutorials. Maybe some general framework to help them, like a template with Docsify....

That being said, if we can easily modify this tutorial and get more consistant feeling, I think it is probably a positive thing.

File: docs/tutorials/basic-dapp/part-4-use-external-accounts.md [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
```
Comment (joepetrowski) [inline docs/tutorials/basic-dapp/part-4-use-external-accounts.md]:
```suggestion
We will declare our DApp and request access to the external accounts in a `useEffect` React hook that we are now familiar with. We used to call `loadAccounts()` to initialize the keyring. This function also accepts injected accounts as arguments, so all we'll have to do is `loadAccounts(injectedAccounts)`. Quite handy, isn't it?
```

**Comment by jeluard** [comment]
Something that I feel could be useful to detail: clearly state at the beginning of each part what API will be used.

**Comment by Tbaut** [comment]
> Something that I feel could be useful to detail: clearly state at the beginning of each part what API will be used.

Do you see this as a mean to make things clearer? To present a tool/api before using, or is there any other motivation for it?

**Comment by jeluard** [comment]
Right exactly to make things clearer. Some introduction detailing which specific API will be used to solve the issue.

File: docs/tutorials/basic-dapp/part-4-use-external-accounts.md (hunk 13 above)
Comment (Tbaut) [reply docs/tutorials/basic-dapp/part-4-use-external-accounts.md]: fixed, thanks

**Comment by Tbaut** [comment]
Everything should now render nicely, with images.
The tutorial has a sidebar on the left.
It appears in the tuto list under a new "Front-end" section with a logo.

Please :heavy_check_mark: if you're good, I'd like to have actual users try it out and see if it's useful to anyone.
